R/teal_lockfile.R

```diff
@@ -23,92 +44,140 @@
+      ...
+    )
+  })
 
-    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
-    lockfile_task$invoke(close = inherits(old_plan, "sequential"), lockfile_path)
-    logger::log_trace("lockfile creation invoked.")
-  }
+  process$invoke(
+    lockfile_path = lockfile_path,
+    run = renv_snapshot,
+    opts = options(),
+    libpaths = .libPaths(),
+    sysenv = as.list(Sys.getenv()), # normally output is a class of "Dlist"
+    wd = getwd()
+  )
+  logger::log_trace("Lockfile creation started based on { getwd() }.")
+
+  process
 }
 
-create_renv_lockfile <- function(close = FALSE, lockfile_path = NULL) {
-  checkmate::assert_flag(close)
-  checkmate::assert_string(lockfile_path, na.ok = TRUE)
-  promise <- promises::future_promise({
-    # Below we can not use a file created in tempdir() directory.
-    # If a file is created in tempdir(), it gets deleted on 'then(onFulfilled' part.
-    shiny::onStop(function() file.remove(lockfile_path))
+#' @rdname teal_lockfile
+#' @keywords internal
+renv_snapshot <- function(lockfile_path = NULL, opts, sysenv, libpaths, wd) {
+  checkmate::assert_string(lockfile_path)
+  checkmate::assert_list(opts)
+  checkmate::assert_class(sysenv, "list")
+  checkmate::assert_character(libpaths, min.len = 1)
+  checkmate::assert_directory(wd)
 
-    renv_logs <- utils::capture.output(
-      renv::snapshot(
-        lockfile = lockfile_path,
-        prompt = FALSE,
-        force = TRUE
-        # type = is taken from renv::settings$snapshot.type()
-      )
+  # mirai starts in vanilla session in the R.home directory. We need to pass all session related info
+  options(opts)
+  do.call(Sys.setenv, sysenv)
+  .libPaths(libpaths)
+  setwd(wd)
+
+  out <- utils::capture.output(
+    renv <- renv::snapshot(
+      lockfile = lockfile_path,
+      prompt = FALSE,
+      force = TRUE
+      # type is taken from renv::settings$snapshot.type()
     )
-    if (any(grepl("Lockfile written", renv_logs))) {
-      logger::log_trace("lockfile created successfully.")
+  )
+
+  list(
+    out = out,
+    lockfile_path = lockfile_path,
+    length = length(renv$Packages)
+  )
+}
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_tracker <- function(process) {
+  checkmate::assert_class(process, "ExtendedTask", null.ok = TRUE)
+  if (inherits(process, "ExtendedTask")) {
+    tracker <- observeEvent(process$status(), {
+      if (process$status() != "running") {
+        teal_lockfile_handler(process)
+      }
+    })
+  } else {
+    # If it is not mirai proces, then lockfile was provided through options. Skip the process - file is ready.
+    shinyjs::show("teal-lockFile")
+  }
+}
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_handler <- function(process) {
+  # todo: make sure that we catch all possible status outputs
+  #     - normally getwd should be set to the app directory (snapshot based on the app files)
+  #     - what happens if app directory is set to a different location?
```

Comment:
App directory can be specified with `shinyAppDir()`, to point to a directory containing `server.R` or `app.R`. If you run shiny app with `shinyAppDir(path)` the `getwd()` shows `path` which is good, and as expected. 

However in `teal` use case `teal::init` creates `ui` and `server` functions, that can be used in `shinyApp(ui = ui, server = server)`. It is possible that the file that launches `shinyApp(ui = init$ui, server = init$srv)` is not in working directory and the `getwd()` is used and points to something else. Investigating on how to track actual directory with the app.

Reply:
I think we may want to use `rstudioapi::getActiveDocumentContext()$path` which tracks the active document context to figure out the app directory

<img width="575" alt="image" src="https://github.com/user-attachments/assets/39d899d2-4b0a-436b-bb9a-e20f61de4619">

Reply:
I think this is too complicated to get this detected by `teal`. @gogonzo what if we introduce new option `options(teal.app.dir)` that hosts the directory of the teal app. This probably would have more applications in other use cases. This can be set to `getwd()` by default.

Reply:
I think we can give up and try something in the future if this problem occurs. I guess it will be very rare to change working directory especially on deployed app.

Reply:
Cool, this would only require use to orchestra paths in `renv::snapshot()` call. So there is a way to include a custom path in there if there is a need in the futrue https://github.com/insightsengineering/teal/pull/1263/files#diff-c389a151f2967297002c1253ee7819f95d9017f250147356c6e8d6d66165638cR79